core/runtime/src/main/java/io/quarkus/runtime/StartupContext.java

```diff
@@ -97,4 +97,19 @@ public String getCurrentBuildStepName() {
     public void setCurrentBuildStepName(String currentBuildStepName) {
         this.currentBuildStepName = currentBuildStepName;
     }
+
+    private record ShutdownTask(int priority, Runnable action) implements Runnable, Comparable<ShutdownTask> {
+
+        @Override
+        public int compareTo(ShutdownTask task) {
+            // with a regular int compare we'll get the asc ordered queue [1,2,3,4]
+            //  but we want the desc order [4,3,2,1] so that the task with the higher priority is polled first:
```

Comment:
Yeah that needs some javadoc :]

Also... shouldn't it be the opposite for `lastShutdownTasks`... Maybe? Though maybe your change is making that concept obsolete, I don't know.